polkadot/xcm/xcm-executor/src/traits/controller.rs

```diff
@@ -0,0 +1,206 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A set of traits that define how a pallet interface with XCM.
+
+use crate::{
+	traits::{QueryHandler, QueryResponseStatus},
+	InteriorMultiLocation,
+	Junctions::Here,
+	MultiLocation, Xcm, Outcome, XcmError
+};
+use frame_support::{pallet_prelude::*, parameter_types};
+use sp_weights::Weight;
+use xcm::{v3::XcmHash, VersionedMultiLocation, VersionedXcm};
+
+/// Umbrella trait for all Controller traits.
+pub trait Controller<Origin, RuntimeCall, Timeout>:
+	ExecuteController<Origin, RuntimeCall> + SendController<Origin> + QueryController<Origin, Timeout>
+{
+}
+
+impl<T, Origin, RuntimeCall, Timeout> Controller<Origin, RuntimeCall, Timeout> for T where
+	T: ExecuteController<Origin, RuntimeCall>
+		+ SendController<Origin>
+		+ QueryController<Origin, Timeout>
+{
+}
+
+/// Weight functions needed for [`ExecuteController`].
+pub trait ExecuteControllerWeightInfo {
+	/// Weight for [`ExecuteController::execute`]
+	fn execute() -> Weight;
+}
+
+/// Execute an XCM locally, for a given origin.
+pub trait ExecuteController<Origin, RuntimeCall> {
+	/// Weight information for ExecuteController functions.
+	type WeightInfo: ExecuteControllerWeightInfo;
+
+	/// Attempt to execute an XCM locally, and return the outcome.
+	///
+	/// # Parameters
+	///
+	/// - `origin`: the origin of the call.
+	/// - `message`: the XCM program to be executed.
+	/// - `max_weight`: the maximum weight that can be consumed by the execution.
+	fn execute(
+		origin: Origin,
+		message: Box<VersionedXcm<RuntimeCall>>,
+		max_weight: Weight,
+	) -> Result<Outcome, DispatchError>;
+}
```

Comment:
We already have the `ExecuteXcm` trait, implemented by the executor, that allows you to execute XCMs locally. Mainly by calling `prepare_and_execute`.

What the `execute` extrinsic in pallet-xcm does is basically convert the origin, do some filtering, versioning, and then call the internal executor, which implements `ExecuteXcm`. I feel like it's confusing to have both `ExecuteXcm` and this `ExecuteController`.

We could name this `ExecuteVersionedXcm`? Or is that even worse? Also I'd add more documentation to the trait to show its difference to the `ExecuteXcm` trait. With documentation it might be enough.

I do agree with the existence of the trait, it could help to make other "pallet-xcm"s in things other than FRAME.

Reply:
yeah the way I see it, is that this Executecontroller trait is basically an umbrella traits for these 3
Let me see if I can come up with a better way to extract this
```rust
		type ExecuteXcmOrigin: EnsureOrigin<
			<Self as SysConfig>::RuntimeOrigin,
			Success = MultiLocation,
		>;

		/// Our XCM filter which messages to be executed using `XcmExecutor` must pass.
		type XcmExecuteFilter: Contains<(MultiLocation, Xcm<<Self as Config>::RuntimeCall>)>;

		/// Something to execute an XCM message.
		type XcmExecutor: ExecuteXcm<<Self as Config>::RuntimeCall>;
```

Reply:
> We could name this ExecuteVersionedXcm? Or is that even worse? Also I'd add more documentation to the trait to show its difference to the ExecuteXcm trait. With documentation it might be enough.

Added the details mentioned here in the doc, specifying as you precised that the implementation of the trait should deal with conversion, versioning and calling lower level api.

Not super convinced by `ExecuteVersionedXcm`, as it does not encapsulate the fact that these traits deal with all the low-level details not just versioning. Controller sounds like a better fit to me, but I am open to suggestions.